swarm/src/behaviour.rs

```diff
@@ -299,21 +290,20 @@ pub enum ToSwarm<TOutEvent, TInEvent> {
         event: TInEvent,
     },
 
-    /// Informs the `Swarm` about an address observed by a remote for
-    /// the local node by which the local node is supposedly publicly
-    /// reachable.
+    /// Reports a new candidate for an external address to the [`Swarm`].
     ///
-    /// It is advisable to issue `ReportObservedAddr` actions at a fixed frequency
-    /// per node. This way address information will be more accurate over time
-    /// and individual outliers carry less weight.
-    ReportObservedAddr {
-        /// The observed address of the local node.
-        address: Multiaddr,
-        /// The score to associate with this observation, i.e.
-        /// an indicator for the trusworthiness of this address
-        /// relative to other observed addresses.
-        score: AddressScore,
-    },
+    /// This address could come from a variety of sources:
+    /// - A protocol such as identify obtained it from a remote.
+    /// - The user provided it based on configuration.
+    /// - We made an educated guess based on one of our listen addresses.
+    /// - We established a new relay connection.
```

Comment:
Do we have any mechanism to confirm a relayed connection? I am not aware of a `libp2p-autonat` implementation that would probe relayed addresses. That said, emitting them as `NewExternalAddrCandidate` makes sense, as we can not confirm them. How would `libp2p-identify` make sure to include relayed addresses when advertising the local node's `listenAddrs` (i.e. external addresses)? Would `libp2p-identify` ignore all `FromSwarm::NewExternalAddr` addresses except for the relayed ones?

Reply:
I don't think so but the relay could (should?) emit an event that the relay address is now an external address of ours.

> How would `libp2p-identify` make sure to include relayed addresses when advertising the local node's `listenAddrs` (i.e. external addresses)? Would `libp2p-identify` ignore all `FromSwarm::NewExternalAddr` addresses except for the relayed ones?

Why ignore them? A relayed address is just one more external address, isn't it? It is feasible to use relays to translate between IPv4 and IPv6 for example if node you want to connect to is only on IPv6 and you are only on IPv4.

Reply:
>> Would `libp2p-identify` ignore all `FromSwarm::NewExternalAddr` addresses except for the relayed ones?
>
> Why ignore them?

There seems to be a misunderstanding. I suggest `libp2p-identify` to track all addresses confirmed via `ExternalAddrConfirmed`. That would not include relayed addresses. Though one still needs to advertise ones relayed addresses via `libp2p-identify`. I don't think one should advertise all `NewExternalAddrCandidate`, but instead only the relayed `NewExternalAddrCandidate`s.

Does that make sense?

Reply:
I think we can treat a relayed address as externally reachable without verifying it. We did connect to the relay after all so it is kind of implicit that others will be able to connect to the relay too. Thus, I think the relay behaviour could just issue a `ToSwarm::NewExternalAddr` with the relayed address.